compiler/rustc_mir_transform/src/lower_intrinsics.rs

```diff
@@ -162,6 +163,35 @@ impl<'tcx> MirPass<'tcx> for LowerIntrinsics {
                             terminator.kind = TerminatorKind::Goto { target };
                         }
                     }
+                    sym::option_payload_ptr => {
```

Comment:
Please add a mir-opt test that demonstrates this lowering.  Here's an example: <https://github.com/rust-lang/rust/pull/109035/files#diff-3f1709a807c26d064317578ef79b45f07e05a6e782a26b0df2fad564ce83f127>

(The output files are generated with `./x.py test tests/mir-opt --bless` -- don't try to write them by hand, as that would be a horrible experience.)